### PR TITLE
Use system allocator for containers on macOS (bypassing Unity's allocator)

### PR DESCRIPTION
**Pull request by rnlahaye**
First part of a two-part fix for #2899.

This change causes STL containers (std::vector and the like) to use the native system allocator on macOS instead of the Unity-provided allocator.

It does this by defining an allocator which avoids using of `new` and `delete`, and then creating aliases to `std::vector` and friends in the `principia::std` namespace using the new allocator. The header file containing this change is included via compiler argument, so this change is almost entirely transparent to the rest of the codebase (I had to add full qualification in two places).

### CODE REVIEW

**Comment by pleroy** [comment]
FWIW, this PR generally looks good.  We'll do a more thorough review over the week-end.  Expect nits, but nothing earth-shattering.

File: Makefile [hunk 1]
```diff
@@ -87,6 +87,7 @@ SHARED_ARGS   := \
 	-Wall -Wpedantic                                              \
 	-Wno-char-subscripts                                          \
 	-Wno-gnu-anonymous-struct                                     \
+	-Wno-c99-extensions                                           \
```
Comment (pleroy) [inline Makefile:90]: Just out of curiosity, which warning does this suppress?

File: magic/malloc_allocator.hpp [hunk 2]
```diff
@@ -0,0 +1,110 @@
+// This file contains two things:
```
Comment (pleroy) [inline magic/malloc_allocator.hpp:1]: Among other things, the `magic` directly is going to interact poorly with VS.  I would organize things differently here:

1. A file `base/malloc_allocator.hpp` that declares and defines the class `MallocAllocator` in namespace `principia::base`.  This one is not magic in any way, and should be included in the usual manner.
2. A file `base/macos_allocator_replacement.hpp` that has all the stuff declared in `principia::std`.  This is the one that is magically included.  For safety, it should have preprocessor directives that ensure that it's never pulled for other platforms:
```c++
#if !OS_MACOSX
#error Only include for macOS
#endif 
... all the rest ...
```
You'll need to include `base/macros.hpp` in that second file.

File: magic/malloc_allocator.hpp [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+template <typename Key, typename Compare = std::less<Key>,
+          typename Allocator = allocator<Key>>
+using multiset = ::std::multiset<Key, Compare, Allocator>;
+
+template <typename Key, typename T, typename Compare = std::less<Key>,
+          typename Allocator = allocator<std::pair<const Key, T>>>
+using multimap = ::std::multimap<Key, T, Compare, Allocator>;
+
+}  // namespace std
+
```
Comment (pleroy) [inline magic/malloc_allocator.hpp:109]: Remove blank line.

Comment (rnlahaye) [reply magic/malloc_allocator.hpp]: Done.

File: magic/malloc_allocator.hpp (hunk 2 above)
Comment (rnlahaye) [reply magic/malloc_allocator.hpp]: Done.

File: Makefile (hunk 1 above)
Comment (rnlahaye) [reply Makefile:90]:
```
./astronomy/orbital_elements_body.hpp:165:10: warning: designated initializers
      are a C99 feature [-Wc99-extensions]
        {.t = time,
         ^~~~~~~~~
./astronomy/orbital_elements_body.hpp:166:10: warning: designated initializers
      are a C99 feature [-Wc99-extensions]
         .a = *osculating_elements.semimajor_axis,
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./astronomy/orbital_elements_body.hpp:167:10: warning: designated initializers
      are a C99 feature [-Wc99-extensions]
         .h = e * Sin(ϖ),
         ^~~~~~~~~~~~~~~
```
…and so on.

Comment (eggrobin) [reply Makefile:90]: Ah, right, these are C++20 whereas the flags say 1z=17.

**Review by eggrobin — APPROVED** [review]

**Pull request merged**